Assets/Implementation/Pathfinding/Scripts/NPCController.cs

```diff
@@ -0,0 +1,213 @@
+{
+public class NPCController : MonoBehaviour
+{
+    [Header("Dependecies")]
+    [SerializeField] private Grid _grid;
+    [SerializeField] private bool _drawGizmos;
+
+    [Header("Movement variables")]
+    [SerializeField] float _moveSpeed = 1f;
+
+    Node[,,] _nodeGrid;
+    List<Node> _backtrackedPath = new List<Node>();
+    [SerializeField] Vector3Int _endNode;
+    int _gridWidth;
+    int _gridHeight;
+    [SerializeField] bool _pathCompleted = true;
+    [SerializeField] Node[] open;
+    [SerializeField] Node[] closed;
+
+    public bool PathCompleted => _pathCompleted;
+
+    private void Start()
+    {
+        _gridWidth = _grid.GridSize.x;
+        _gridHeight = _grid.GridSize.y;
+    }
+
+    public Vector3Int GetGridPosition()
+    {
+        var pos = transform.position;
+
+        //clamp to grid positions
+        var clampedValue = new Vector2(RoundToMultiple(pos.x, _grid.CellSize),
+            RoundToMultiple(pos.y, _grid.CellSize));
+
+        return new Vector3Int((int)clampedValue.x, (int)clampedValue.y, 0);
+    }
+
+    void Update()
+    {
+        var pos = Camera.main.ScreenToWorldPoint(Input.mousePosition);
+
+        //clamp to grid positions
+        var clampedValue = new Vector2(RoundToMultiple(pos.x, _grid.CellSize),
+            RoundToMultiple(pos.y, _grid.CellSize));
+
+        if (Input.GetKeyDown(KeyCode.P) & _pathCompleted)
+        {
+            CreateGrid();
+            StopAllCoroutines();
+
+            _backtrackedPath.Clear();
+
+            _endNode = new Vector3Int((int)clampedValue.x, (int)clampedValue.y, 0);
+
+            FindPath(_endNode);
+            StartCoroutine(MoveToTarget(_backtrackedPath));
+        }
+    }
+
+    public void SetTarget(Vector3Int position)
+    {
+        //clamp to grid positions
+        var clampedValue = new Vector2(RoundToMultiple(position.x, _grid.CellSize),
+            RoundToMultiple(position.y, _grid.CellSize));
+
+        CreateGrid();
+        StopAllCoroutines();
+
+        _backtrackedPath.Clear();
+
+        _endNode = new Vector3Int((int)clampedValue.x, (int)clampedValue.y, 0);
+
+        FindPath(_endNode);
+        StartCoroutine(MoveToTarget(_backtrackedPath));
+    }
+
+    /// <summary>
+    /// Move the NPC to the target (This is just for visualisation purposes, needs to be replanced with actual movement system)
+    /// </summary>
+    private IEnumerator MoveToTarget(List<Node> path)
+    {
+        _pathCompleted = false;
+        for (int i = path.Count - 1; i >= 0; i--)
+        {
+            while (Vector3.Distance(path[i].position, transform.position) > 0.1f)
+            {
+                Vector3 direction = path[i].position - transform.position;
+                transform.position += direction.normalized * _moveSpeed * Time.deltaTime;
+                yield return new WaitForEndOfFrame();
+            }
+            transform.position = path[i].position;
+        }
+        _pathCompleted = true;
```

Comment:
Think it's cooler and better to use a callback that executes when the path is complete. This way, you don't have to constantly poll whether the path is done. Instead, the function is executed when the path is completed